cmd/apmtool/sourcemap.go

```diff
@@ -0,0 +1,115 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"os"
+
+	"github.com/urfave/cli/v3"
+)
+
+func (cmd *Commands) uploadSourcemapCommand(c *cli.Context) error {
+	var data bytes.Buffer
+	mw := multipart.NewWriter(&data)
+	mw.WriteField("service_name", c.String("service-name"))
+	mw.WriteField("service_version", c.String("service-version"))
+	mw.WriteField("bundle_filepath", c.String("bundle-filepath"))
+	sourcemapFileWriter, err := mw.CreateFormFile("sourcemap", "sourcemap.js.map")
+	if err != nil {
+		return err
+	}
+	if filename := c.String("file"); filename == "-" {
+		if _, err := io.Copy(sourcemapFileWriter, os.Stdin); err != nil {
+			return err
+		}
+	} else {
+		f, err := os.Open(filename)
+		if err != nil {
+			return fmt.Errorf("error opening file: %w", err)
+		}
+		if _, err := io.Copy(sourcemapFileWriter, f); err != nil {
+			f.Close()
+			return err
+		}
+		f.Close()
+	}
+	if err := mw.Close(); err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(
+		http.MethodPost,
+		cmd.cfg.KibanaURL+"/api/apm/sourcemaps",
+		bytes.NewReader(data.Bytes()),
+	)
+	if err != nil {
+		return fmt.Errorf("error creating HTTP request: %w", err)
+	}
+	req.SetBasicAuth(cmd.cfg.Username, cmd.cfg.Password)
+	req.Header.Set("Content-Type", mw.FormDataContentType())
+	req.Header.Set("kbn-xsrf", "1")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("error performing HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+	io.Copy(os.Stderr, resp.Body)
+	fmt.Fprintln(os.Stderr)
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("error uploading sourcemap; server responded with %q", resp.Status)
+	}
+	return nil
+}
+
+// NewUploadSourcemapCmd returns pointer to a Command that uploads a source map to Kibana
+func NewUploadSourcemapCmd(commands *Commands) *cli.Command {
+	return &cli.Command{
+		Name:   "upload-sourcemap",
+		Usage:  "upload a source map to Kibana",
+		Action: commands.uploadSourcemapCommand,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:     "file",
+				Aliases:  []string{"f"},
+				Required: true,
+				Usage:    "File containing the sourcemap to upload. Use '-' to read from stdin.",
```

Comment:
there'll be two ways to trigger stdin reading, either specifying `""` (espoll command) or `-` (this command). We should settle for one.

Reply:
We do also have the same usage as above for the `send-events` command. As `-` is used in more than one place already, how about we settle for `-` instead of `""`?

Reply:
On second thought, it doesn't really make sense to expect `-` value for the `query` flag and parse stdin whereas it makes sense for the file flag.